### ISSUE TITLE
Java API for writing macros.

### ISSUE THREAD

**Issue by adventuremagic123** [RPTools/maptool#1910]
**Is your feature request related to a problem? Please describe.**
Some of my macros are complex and take a lot of time for MapTool to interpret -- and then execute.  I've done a test where I capture the start and stop time, and it turns out most of the time was spent in the interpreter interpreting the code.

**Describe the solution you'd like**
I've actually used .NET, but since MapTool appears to be developed using Java, I'm asking for a Java API of some sort that would enable us to write macros using Java.  I suspect that compiled Java code will run faster than interpreted MTScript.

**Describe alternatives you've considered**
I'm trying to use caching to mitigate the performance problem, but since sometimes the issue is just the amount of time it takes to interpret the code in the interpreter -- there's really nothing I can do except write simpler code.  That limits what I can do with the macros.

**Additional context**
None.

**Comment by nmeier** [comment]
could you share some of your test macros? I'm trying to tune the macro performance and need test cases.

**Comment by adventuremagic123** [comment]
Really, most of my problem appears to be the amount of time spent in the interpreter.  You can see this by capturing the start and stop time in the macro -- and timing the whole execution with a stopwatch.  You'll find that for some complex macros or perhaps macros structured a certain way that the stopwatch will show more time consumed than the start and stop time captured by the macro.  The difference is the time spent in the interpreter.

**Comment by adventuremagic123** [comment]
I don't have anything for test cases.  Everything I did was temporary as I investigated performance issues.

**Comment by adventuremagic123** [comment]
It's important to note that currently with MTScript that I do NOT use the embedded HTML feature.  Instead, I always build my HTML as strings and call broadcast() so that I can precisely control who sees the HTML.

That technique should work great with a Java API.

**Comment by nmeier** [comment]
basically related to #1898

once the Parser and Expressions are created the scripts are pretty fast. Right now there's no real optimization though: they are instantiated for every roll element, lists and arrays are not handled well. 

So a plugin mechanism is a way for a real extension - as in: plug in a jar, get MapToolPossfinder or whatever. Having external code link deep into the spaghetti of MapTool could lead to the need to update plugins on every major release. Food for thought.

**Comment by adventuremagic123** [comment]
Yes, once the parsing phase is completed, the scripts run really fast.

If the performance of the parsing phase could be greatly improved, that might be enough to resolve this issue.

If the real answer is that a Java API should be developed, then it should be done in a way (like MTScript) that insulates API users from MapTool changes and internal structure.

**Comment by JamzTheMan** [comment]
To bad you couldn't parse and "compile" each macro when you save it so it has less to do when you run it over and over...

**Comment by nmeier** [comment]
I have some changes pending that compiles each macro line only once at least. Currently a loop like this compiles the 'do something else' for each loop iteration. That's terribly slow.

```
[h, FOREACH(i, is), code:{
	[h: do something else] 
}]

```
